PERF: sparse to_csv

rtlee9 opened the pull request:
Improves `NDFrame.to_csv` performance for sparse dataframe by casting to dense before initializing `DataFrameFormatter`. Results in many fewer calls to `to_native_types` which saves time. Added a new ASV benchmark based on the example provided by OP in #41023.

Benchmark results:
```
[100.00%] ··· io.csv.ToCSVSparse.time_sparse_to_dense_to_csv                                        1.16±0.02s       before           after         ratio
     [56d82a9b]       [945a3525]
     <main~1>         <main>
-       13.6±0.1s       1.22±0.04s     0.09  io.csv.ToCSVSparse.time_sparse_to_csv

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

- [x] closes #41023 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

rtlee9 committed:
PERF: sparse to_csv

Improves to_csv performance for sparse matric by casting to dense
before initializing DataFrameFormatter. Results in many fewer calls to
`to_native_types` which saves time.

phofl commented on pandas/core/generic.py:
```diff
@@ -3717,6 +3717,8 @@ def to_csv(
         >>> df.to_csv('folder/subfolder/out.csv')  # doctest: +SKIP
         """
         df = self if isinstance(self, ABCDataFrame) else self.to_frame()
+        if hasattr(df, "sparse"):
```
Is it possible to fix this later?

rtlee9 replied:
Fix in what way? Open to suggestions

phofl replied:
Deeper in the call stack. We are formatting the values in the end before writing them to the file. Maybe there

rtlee9 replied:
Moved it a few calls deeper to `CSVFormatter.__init__`

rtlee9 committed:
Move optimization deeper in the call stack

rtlee9 committed:
Merge remote-tracking branch 'upstream/main'

mzeitlin11 commented on doc/source/whatsnew/v2.0.0.rst:
```diff
@@ -163,6 +163,7 @@ Performance improvements
 - Performance improvements to :func:`read_sas` (:issue:`47403`, :issue:`47405`, :issue:`47656`, :issue:`48502`)
 - Memory improvement in :meth:`RangeIndex.sort_values` (:issue:`48801`)
 - Performance improvement in :class:`DataFrameGroupBy` and :class:`SeriesGroupBy` when ``by`` is a categorical type and ``sort=False`` (:issue:`48976`)
+- Performance improvement for saving to CSV with :class:`CSVFormatter` when data frame is sparse (:issue:`41023`)
```
Can you mention the user-facing `to_csv` instead of `CSVFormatter`?

mzeitlin11 commented on pandas/io/formats/csvs.py:
```diff
@@ -69,7 +69,10 @@ def __init__(
         self.fmt = formatter
 
-        self.obj = self.fmt.frame
+        if hasattr(self.fmt.frame, "sparse"):
```
Does this only catch cases where the entire dataframe is sparse? What about when only some columns are `SparseArray`?

I worry a bit here about the forced materialization of the full sparse data at once. Since formatting occurs column-by-column, would it be possible to instead materialize each _column_ as necessary instead of the entire frame at once? Similarly, this has the downside of materializing the full dataframe without regard for `chunksize`. `chunksize` will always make serialization slower, but not respecting it (along with column issue) might cause problematically higher memory usage.

mzeitlin11 replied:
Is it possible to instead handle this later by adjusting `_format_native_types`?

rtlee9 replied:
Based on my [profiling](https://gist.github.com/rtlee9/bf5cb69494532337f9533598624a7a5d) `_format_native_types` did not account for any of the difference in performance between forced materialization (fast) and the current behavior (slow) so wouldn't help as a result. The same profiler [showed](https://gist.github.com/rtlee9/ca34f6693312c3f5122753b703cf2c2f) that `to_native_types`, which is called earlier in `_save_chunk`, was more of a bottleneck due to number of calls rather than time / call.  Looks like this happens because there are many more blocks without the forced materialization (maybe one per row).


We could potentially implement this logic deeper in the call stack such as in `managers.to_native_types` but I'm not familiar with this class and it seemed like it could have broader-than-intended effects.

rtlee9 replied:
I'm not sure I follow what you're saying about chunksize. AFAIK this PR doesn't change the chunking behavior in `_save_body` in that `chunksize` is unchanged and there are the same number of calls to `_save_chunk`. I do understand the idea that materializing sparse data can consume more memory but still trying to connect this with chunksize in my mind.

For a dataframe with mixed sparse and dense columns you're right that this PR won't do anything. `hasattr(pd.DataFrame({"A": pd.arrays.SparseArray([0, 1]), "B": [0, 0]}), 'sparse')` evaluates to false. Seems like materializing column-by-column could help with this particular case but is it worth the complexity if that is the only benefit (i.e., is this a common scenario)?

mzeitlin11 replied:
> I'm not sure I follow what you're saying about chunksize. AFAIK this PR doesn't change the chunking behavior in `_save_body` in that `chunksize` is unchanged and there are the same number of calls to `_save_chunk`. I do understand the idea that materializing sparse data can consume more memory but still trying to connect this with chunksize in my mind.
> 
The way I thought this worked (could be wrong :) was essentially the following psuedocode:
```
for chunk in chunks
   do_format(chunk)  # Causes sparse to dense conversion
   ...
```
So the main concern is that this pr causes a change from `1 chunksize of data converted to dense at a time` -> `whole data converted to dense at once`. So while an awesome perf improvement, does this cause problems for sparse data that can't be fit into memory?

phofl replied:
Yes, agreed. We should materialise column by column. This is what I referenced initially

rtlee9 replied:
Ok so it sounds like there are two concerns:
1. Materializing the entire frame at once rather than in chunks can increase memory consumption drastically
2. This pr does not currently affect frames that contain some sparse columns and some dense columns

For 1) this makes a lot of sense that this could be an issue and 6d092ee6 addresses this by moving the materialization to `_save_chunk`. Unsurprisingly this erases some of the performance gains (see comment below) but I agree this is worth the memory preservation. In the future maybe we can do something dynamic where we optionally convert the whole frame to dense all at once to reap the additional perf gains but only if sufficient memory (or if the user specifies).

For 2) I'm not sure the best way to implement column-wise materialization while preserving the perf gains. A simple iteration over columns and casting to_dense results in the same number of blocks and no perf gains, so would need to do something beyond this.

phofl replied:
This should be done in ``to_native_types`` I think, this is in a different place now, but has the same issues as before

rtlee9 replied:
It resolves the first of the two issues I outlined above — it materializes within chunks to save memory

mzeitlin11 commented on asv_bench/benchmarks/io/csv.py:
```diff
@@ -172,6 +172,26 @@ def time_head_of_multiindex(self):
+        self.data.to_csv("sparse_pd.csv")
+
+    def time_sparse_to_dense_to_csv(self):
+        self.data.sparse.to_dense().to_csv("sparse_pd.csv")
```
Can you also please shrink the size of the data for the benchmark? I think we usually target < 100ms, but based on PR description looks like this is much slower (especially before this commit)

mzeitlin11 commented on asv_bench/benchmarks/io/csv.py:
```diff
@@ -172,6 +172,26 @@ def time_head_of_multiindex(self):
+        self.data = DataFrame.sparse.from_spmatrix(sparse_mtx)
+
+    def time_sparse_to_csv(self):
+        self.data.to_csv("sparse_pd.csv")
```
Following the pattern of the benchmarks above, should this be `self.fname` instead?

rtlee9 replied:
Yeah good call — updated in [6fa18f6](https://github.com/pandas-dev/pandas/pull/49066/commits/6fa18f642e3a242d3d716a322d6c51d11bb008c4)

rtlee9 committed:
Update whatsnew to reference user-facing `to_csv` instead of `CSVFormatter`

rtlee9 committed:
Reduce test frame size and use self.fname for asv test case

rtlee9 committed:
Merge remote-tracking branch 'upstream/main'

rtlee9 committed:
Move sparse conversion deeper in the call stack

This should improve memory consumption by only materializing one chunk
at a time

rtlee9 committed:
Merge remote-tracking branch 'upstream/main'

rtlee9 commented:
Revised asv benchmarks (upstream/main vs pr) after moving materialization to the chunk level to preserve memory. Chunk-level materialization takes longer than all-at-once materialization but is still a significant improvement over upstream/main.

```
[100.00%] ··· io.csv.ToCSVSparse.time_sparse_to_dense_to_csv                                         87.2±3ms
       before           after         ratio
     [9c9789c5]       [6d092ee6]
     <main^2>         <main>
-         641±4ms          285±3ms     0.44  io.csv.ToCSVSparse.time_sparse_to_csv

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

phofl commented:
Please don't wait for reviewers to resolve conversations

rtlee9 committed:
Merge remote-tracking branch 'upstream/main'

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
